src/plugins/owner/MultisigPlugin.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.25;
+
+import {PackedUserOperation} from "@eth-infinitism/account-abstraction/interfaces/PackedUserOperation.sol";
+
+import {FunctionReference, FunctionReferenceLib} from "../../helpers/FunctionReferenceLib.sol";
+import {_coalescePreValidation} from "../../helpers/ValidationDataHelpers.sol";
+import {IPlugin} from "../../interfaces/IPlugin.sol";
+import {IValidation} from "../../interfaces/IValidation.sol";
+import {BasePlugin} from "../BasePlugin.sol";
+import {PluginManifest, PluginMetadata} from "../../interfaces/IPlugin.sol";
+
+// Non-threshold based multisig plugin - all owners must sign.
+// Supports up to 100 owners per id.
+contract MultisigPlugin is IValidation, BasePlugin {
+    struct OwnerInfo {
+        uint256 length;
+        FunctionReference[100] validations;
+    }
+
+    uint256 internal constant _SIG_VALIDATION_PASSED = 0;
+    uint256 internal constant _SIG_VALIDATION_FAILED = 1;
+
+    mapping(uint8 id => mapping(address account => OwnerInfo)) public ownerInfo;
+
+    error AlreadyInitialized();
+    error ArrayLengthMismatch();
+    error NotAuthorized();
+    error NotInitialized();
+    error InvalidOwners();
+    error ValidationReturnedAuthorizer();
+
+    /// @inheritdoc IPlugin
+    function onInstall(bytes calldata data) external override {
+        uint8 id = uint8(bytes1(data[:1]));
+
+        if (ownerInfo[id][msg.sender].length != 0) {
+            revert AlreadyInitialized();
+        }
+
+        FunctionReference[] memory validations = abi.decode(data[1:], (FunctionReference[]));
+
+        if (validations.length == 0 || validations.length > 100) {
+            revert InvalidOwners();
+        }
+
+        ownerInfo[id][msg.sender].length = validations.length;
+
+        for (uint256 i = 0; i < validations.length; i++) {
+            ownerInfo[id][msg.sender].validations[i] = validations[i];
+        }
+    }
+
+    /// @inheritdoc IPlugin
+    function onUninstall(bytes calldata data) external override {
+        uint8 id = uint8(bytes1(data[:1]));
+
+        uint256 length = ownerInfo[id][msg.sender].length;
+
+        if (length == 0) {
+            revert NotInitialized();
+        }
+
+        for (uint256 i = 0; i < length; i++) {
+            ownerInfo[id][msg.sender].validations[i] = FunctionReference.wrap(bytes21(0));
+        }
+
+        ownerInfo[id][msg.sender].length = 0;
+    }
+
+    /// @inheritdoc IValidation
+    function validateUserOp(uint8 functionId, PackedUserOperation calldata userOp, bytes32 userOpHash)
+        external
+        override
+        returns (uint256)
+    {
+        OwnerInfo storage info = ownerInfo[functionId][userOp.sender];
```

Comment:
Actually, wouldn't this violate the storage access rules of 4337?

Reply:
The account address is the innermost mapping lookup, so the storage should be associated. If we did `ownerInfo[userOp.sender][functionId]` it would not be associated.

As for switching from `msg.sender` to `userOp.sender`, It should be OK as long as the account, and any validation functions installed on the account, choose to send the correct account address during validation. When going from the account to the outermost validation function, it will always be the correct address, and each inner validation function call needs to also propagate the correct address. If it doesn't, then it would be a bug in the validation plugin that makes it not usable via a regular bundler.

Reply:
>If we did ownerInfo[userOp.sender][functionId] it would not be associated.

Is this a quirk? How would switching the order make a difference?
`ownerInfo[userOp.sender][functionId]` => `keccak256(abi.encode(functionId, keccak256(abi.encode(userOp.sender, p))))`
v.s.
`ownerInfo[functionId][userOp.sender]` => `keccak256(abi.encode(userOp.sender, keccak256(abi.encode(functionId, p))))`

[The docs](https://www.erc4337.io/docs/bundlers/running-a-bundler#storage-associated-with-an-address) are not very clear. Based on what they say there, both above should be valid since the slot is uniquely related this address.

Reply:
The docs state that the following is one way to get an associated storage slot:
> `keccak256(A || X) + n`

Where A is the associated address, X is any value, and n is a number between 0 and 128. By solidity's convention, X will be the "virtual" storage slot of the mapping, which doesn't actually hold anything.

In the first case, where it is not associated, the hashing to get the slot is:
```solidity
keccak256(abi.encode(functionId, keccak256(abi.encode(userOp.sender, p))))
```
The function ID is not the associated address, so the output of this expression will not be an associated storage slot. The address gets included in the nested `keccak256` call, but that expression becomes X, not A, in the associated storage calculation.

I found this resource to be helpful in understanding solidity's builtin calculation of mapping storage slots: https://programtheblockchain.com/posts/2018/03/09/understanding-ethereum-smart-contract-storage/

Reply:
Missed `keccak256(A || X) + n` from the docs. 
Good catch `The address gets included in the nested keccak256 call, but that expression becomes X, not A, in the associated storage calculation.`.

Have any idea why `keccak256(X || f(A)) + n` not included? If the goal is uniqueness, it satisfies. Might be a bit hard to enforce the rule in the client/bundler.

Reply:
I think if you include both, then it loses uniqueness. For example, a `mapping(address outer => mapping(address inner => uint256)) value;` would be associated with both the outer and inner addresses. This happens with a lot of ERC-20 allowances - by default, the OZ ERC20 contract will have allowances associated with the spender, instead of the token holder.

As for why they didn't choose `keccak256(X || f(A)) + n` instead of the actual, I think it's just easier on the tracer code within the bundler to detect. But it does make it more annoying for solidity devs, lol.